File: backend/app/execution/capabilities/pareto.py
"""pareto — 80/20 analysis: sort by value desc + cumulative % + optional chart."""

from __future__ import annotations

import re
from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range
from app.execution.utils import parse_number_flexible


def _col_letters_to_index(s: str) -> int:
    n = 0
    for ch in s:
        n = n * 26 + (ord(ch) - ord("A") + 1)
    return n - 1


def _index_to_col_letters(i: int) -> str:
    n = i + 1
    out = ""
    while n > 0:
        rem = (n - 1) % 26
        out = chr(ord("A") + rem) + out
        n = (n - 1) // 26
    return out


def _ensure_2d(v: Any) -> list[list[Any]]:
    if v is None:
        return []
    if not isinstance(v, list):
        return [[v]]
    if not v:
        return []
    if not isinstance(v[0], list):
        return [v] if len(v) > 1 else [[v[0]]]
    return v


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    data_range = params.get("dataRange")
    output_range = params.get("outputRange")
    include_chart = bool(params.get("includeChart", True))
    has_headers = bool(params.get("hasHeaders", True))

    if not data_range or not output_range:
        return {"status": "error", "message": "pareto requires 'dataRange' and 'outputRange'."}

    src = resolve_range(ctx.workbook_handle, data_range)
    src_used = src.current_region if src.count > 1 else src

    if src_used.columns.count < 2:
        return {"status": "error", "message": "dataRange must have 2 columns (label | value)."}

    raw = _ensure_2d(src_used.value)
    start = 1 if has_headers else 0
    rows = []
    for r in raw[start:]:
        lbl = r[0]
        val = r[1] if len(r) > 1 else None
        # Tolerate text-stored numbers from CSV imports ($1,234, 50%, etc.)
        num = parse_number_flexible(val)
        if num is not None:
            rows.append({"label": lbl if lbl is not None else "", "value": num})
    if not rows:
        return {"status": "error", "message": "No numeric value rows found."}

    rows.sort(key=lambda r: r["value"], reverse=True)
    total = sum(r["value"] for r in rows)
    running = 0.0
    output: list[list[Any]] = [["Label", "Value", "Cumulative %"]]
    for r in rows:
        running += r["value"]
        output.append([r["label"], r["value"], 0 if total == 0 else running / total])

    if ctx.dry_run:
        return {"status": "preview", "message": f"Would build Pareto with {len(rows)} item(s)."}

    out = resolve_range(ctx.workbook_handle, output_range)
    out_top = out[0, 0]
    out_sheet = out.sheet
    out_addr_tail = out.address.split("!")[-1].split(":")[0]
    m = re.match(r"^\$?([A-Z]+)\$?(\d+)$", out_addr_tail)
    if not m:
        return {"status": "error", "message": f"Could not parse output address: {out.address}"}
    out_col_l = m.group(1)
    out_col_idx = _col_letters_to_index(out_col_l)
    out_row = int(m.group(2))

    try:
        block_target = out_sheet.range(
            (out_row, out_col_idx + 1),
            (out_row + len(output) - 1, out_col_idx + 3),
        )
        block_target.value = output

        # Format cumulative-% column as 0.0%.
        pct_col_l = _index_to_col_letters(out_col_idx + 2)
        out_sheet.range(f"{pct_col_l}{out_row + 1}:{pct_col_l}{out_row + len(output) - 1}").number_format = "0.0%"
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Write failed: {exc}", "error": str(exc)}

    chart_name: str | None = None
    if include_chart:
        try:
            chart_range = out_sheet.range(
                f"{out_col_l}{out_row}:{_index_to_col_letters(out_col_idx + 2)}{out_row + len(output) - 1}"
            )
            chart = out_sheet.charts.add()
            chart.chart_type = "column_clustered"
            chart.set_source_data(chart_range)
            try:
                chart.api[1].ChartTitle.Text = "Pareto Analysis"
            except Exception:
                pass
            chart_name = chart.name
        except Exception:
            pass

    output_addr = f"{out_sheet.name}!{out_col_l}{out_row}:{_index_to_col_letters(out_col_idx + 2)}{out_row + len(output) - 1}"
    return {
        "status": "success",
        "message": f"Pareto written with {len(rows)} item(s). Output: {output_addr}.",
        "outputs": {"outputRange": output_addr, **({"chartName": chart_name} if chart_name else {})},
    }


registry.register("pareto", handler, mutates=True)

File: frontend/src/engine/capabilities/pareto.ts
/**
 * pareto — 80/20 analysis.
 *
 * Input: 2-col dataRange (label | value). Output: 3 columns (label | value |
 * cumulative %) with rows sorted by value descending. Optionally a combo
 * chart (column for value, line for cumulative %) — via two stacked chart
 * series since Office.js "combo" chart type creation is inconsistent
 * across versions.
 */

import { CapabilityMeta, ParetoParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";
import { parseNumberFlexible } from "../utils/parseNumberFlexible";

const meta: CapabilityMeta = {
  action: "pareto",
  description: "Pareto (80/20) analysis: sorted values + cumulative % + optional chart",
  mutates: true,
  affectsFormatting: false,
  requiresApiSet: "ExcelApi 1.2",
};

function indexToLetters(idx: number): string {
  let n = idx + 1;
  let out = "";
  while (n > 0) {
    const rem = (n - 1) % 26;
    out = String.fromCharCode(65 + rem) + out;
    n = Math.floor((n - 1) / 26);
  }
  return out;
}
function columnLetterToIndex(letters: string): number {
  let n = 0;
  for (let i = 0; i < letters.length; i++) n = n * 26 + (letters.charCodeAt(i) - 64);
  return n - 1;
}

async function handler(
  context: Excel.RequestContext,
  params: ParetoParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { dataRange, outputRange, includeChart = true, hasHeaders = true } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would build Pareto from ${dataRange}` };
  }

  options.onProgress?.("Computing Pareto (sort + cumulative %)...");

  const src = resolveRange(context, dataRange);
  const srcUsed = src.getUsedRange(false);
  srcUsed.load(["values", "rowCount", "columnCount", "address"]);
  await context.sync();

  if (srcUsed.columnCount < 2) {
    return { stepId: "", status: "error", message: "dataRange must have 2 columns (label | value)." };
  }

  const values = (srcUsed.values ?? []) as (string | number | boolean | null)[][];
  const startRow = hasHeaders ? 1 : 0;
  const rows: { label: string | number; value: number }[] = [];
  for (let r = startRow; r < values.length; r++) {
    const lbl = values[r][0];
    // parseNumberFlexible: handles text-stored numbers, currency prefixes,
    // percent, EU format, paren-negatives — anything a CSV import might have.
    const v = parseNumberFlexible(values[r][1]);
    if (v !== null) {
      rows.push({ label: (lbl ?? "") as string | number, value: v });
    }
  }
  if (rows.length === 0) {
    return { stepId: "", status: "error", message: "No numeric value rows found." };
  }

  // Sort descending by value.
  rows.sort((a, b) => b.value - a.value);
  const total = rows.reduce((sum, r) => sum + r.value, 0);
  let running = 0;
  const output: (string | number)[][] = [["Label", "Value", "Cumulative %"]];
  for (const r of rows) {
    running += r.value;
    output.push([r.label, r.value, total === 0 ? 0 : running / total]);
  }

  const out = resolveRange(context, outputRange);
  out.load(["address", "worksheet/name"]);
  await context.sync();
  const outAddrPart = out.address.includes("!") ? out.address.split("!").pop()! : out.address;
  const outFirst = outAddrPart.split(":")[0].match(/^([A-Z]+)(\d+)$/);
  if (!outFirst) {
    return { stepId: "", status: "error", message: `Could not parse output address: ${out.address}` };
  }
  const outStartCol = outFirst[1];
  const outStartColIdx = columnLetterToIndex(outStartCol);
  const outStartRow = Number(outFirst[2]);
  const outSheet = out.worksheet;

  try {
    const block = outSheet.getRangeByIndexes(outStartRow - 1, outStartColIdx, output.length, 3);
    block.values = output as unknown as (string | number | boolean)[][];
    // Format the cumulative-% column.
    const pctCol = indexToLetters(outStartColIdx + 2);
    const pctRange = outSheet.getRange(
      `${pctCol}${outStartRow + 1}:${pctCol}${outStartRow + output.length - 1}`,
    );
    pctRange.numberFormat = [["0.0%"]] as unknown as string[][];
    await context.sync();
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    return { stepId: "", status: "error", message: `Failed to write Pareto block: ${msg}`, error: msg };
  }

  let chartName: string | undefined;
  if (includeChart) {
    try {
      const chartRange = outSheet.getRange(
        `${outStartCol}${outStartRow}:${indexToLetters(outStartColIdx + 2)}${outStartRow + output.length - 1}`,
      );
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const chart: any = (outSheet as any).charts.add("columnClustered", chartRange, "auto");
      chart.title.text = "Pareto Analysis";
      chartName = chart.name ?? "Pareto";
      await context.sync();
    } catch {
      // Chart failure shouldn't abort.
    }
  }

  const outputAddr = `${outSheet.name}!${outStartCol}${outStartRow}:${indexToLetters(outStartColIdx + 2)}${outStartRow + output.length - 1}`;
  return {
    stepId: "",
    status: "success",
    message: `Pareto analysis written with ${rows.length} item(s). Output: ${outputAddr}.${chartName ? ` Chart: ${chartName}.` : ""}`,
    outputs: { outputRange: outputAddr, ...(chartName ? { chartName } : {}) },
  };
}

registry.register(meta, handler as any);
export { meta };
